File: Codes in C++/Dynamic Programming/Decode Ways.cpp
    int numDecodings(string s) {
        int n = s.size();
        vector<int> dp(n + 1, 0);
        dp[n] = 1;
        for (int i = n - 1; i >= 0; i--) {
            if (s[i] == '0') {
                dp[i] = 0;
                continue;
            }
            int res = dp[i + 1];
            if (i + 1 < s.length() &&
                (s[i] == '1' || (s[i] == '2' && s[i + 1] <= '6')))
                res += dp[i + 2];
            dp[i] = res;
        }
        return dp[0];
    }

// class Solution {
// public:
//     int helper(int i,string s,vector<int> &dp){
//         if(i==s.length())
//             return 1;
//         if(dp[i]!=-1) return dp[i];
//         if(s[i]=='0')
//             return 0;
//         int res=helper(i+1,s,dp);
//         if(i+1<s.length()&&(s[i]=='1'||(s[i]=='2'&&s[i+1]<='6')))
//             res+=helper(i+2,s,dp);
//         return dp[i]=res;
//     }
//     int numDecodings(string s) {
//         vector<int> dp(s.length(),-1);
//         return helper(0,s,dp);
//     }
// };

File: Codes in Python3/Dynamic Programming/Decode Ways.py
#DP Bottom-Up Space Omptimized Approach
# Key Observation anything starting from 0 will
# not be part part of any sequence
class Solution:
    def numDecodings(self, s: str) -> int:
        next_one=1 #Corresponds to dp[len(s))
        next_two=0
        
        for i in range(len(s)-1,-1,-1):
            current=0
            if s[i]=="0":
                current=0
            else:
                current=next_one
            if i+1<len(s):
                if (s[i]=="1" or (s[i]=="2" and s[i+1]<="6")):
                    current+=next_two
            next_two=next_one
            next_one=current
            
        return next_one   

# TC=O(n^2)
# Sc=O(1)

#DP Bottom-Up Approach
# Key Observation anything starting from 0 will
# not be part part of any sequence
class Solution:
    def numDecodings(self, s: str) -> int:
        dp = {len(s) : 1}
        
        for i in range(len(s)-1,-1,-1):
            if s[i]=="0":
                dp[i]=0
            else:
                dp[i]=dp[i+1]
            if i+1<len(s):
                if (s[i]=="1" or (s[i]=="2" and s[i+1]<="6")):
                    dp[i]+=dp[i+2]
        return dp[0]   

# TC=O(n^2)
# Sc=O(n)


#DP Recursion Approach
# Key Observation anything starting from 0 will
# not be part part of any sequence
class Solution:
    def numDecodings(self, s: str) -> int:
        dp = {len(s) : 1}
        def dfs(i:int):
            if i in dp:
                return dp[i]
            if s[i]=="0":
                return 0
            dp[i]=dfs(i+1)
            if i+1<len(s):
                if (s[i]=="1" or (s[i]=="2" and s[i+1]<="6")):
                    dp[i]+=dfs(i+2)
            res=dp[i]
            return res
        return dfs(0)
# TC=O(n^2)
# Sc=O(n)

#Recursion Approach
# Key Observation anything starting from 0 will
# not be part part of any sequence
class Solution:
    def numDecodings(self, s: str) -> int:
        def dfs(i:int):
            if i==len(s):
                return 1
            if s[i]=="0":
                return 0
            res=dfs(i+1)
            if i+1<len(s):
                if (s[i]=="1" or (s[i]=="2" and s[i+1]<="6")):
                    res+=dfs(i+2)
            return res
        return dfs(0)
# TC=O(2^n)
# Sc=O(n)